### ISSUE TITLE
replacing a custom label leads to no more log entries

### ISSUE THREAD

**Issue by ctaggart** [GoogleCloudPlatform/google-cloud-dotnet#1036]
When we add a second custom label, the logs stop working.

``` fsharp
    log.Infof "set the batch label"
    log.AddLabel "batch" (sprintf "%d" batchId)
    log.Infof "batch label was set for the first time"
    let batchId = batchId + 1
    log.Infof "set the batch label a second time"
    log.AddLabel "batch" (sprintf "%d" batchId)
    log.Infof "batch label was set a second time"
```

Results in this where the last entry is missing.
![image](https://cloud.githubusercontent.com/assets/80104/25910354/5bd06e32-35b0-11e7-8177-9a331603005b.png)

``` fsharp
let appender = new Google.Cloud.Logging.Log4Net.GoogleStackdriverAppender()
...

        member x.AddLabel key value =
            appender.AddCustomLabel(Google.Cloud.Logging.Log4Net.GoogleStackdriverAppender.Label(Key=key, Value=value))
```

I was expecting `AddLabel` to act like a Dictionary and replace the previous label with the same key.

**Comment by jskeet** [comment]
Any chance you could provide a complete example in C# that would let us reproduce this easily?

**Comment by ctaggart** [comment]
> Any chance you could provide a complete example in C# that would let us reproduce this easily?

@jskeet, I thought I saw on twitter that you were going to work on your F# skills!? :alien:

Yes, will do.

**Comment by ctaggart** [comment]
The code to reproduce is in orphan branch `issue1036` in my fork here:
https://github.com/ctaggart/google-cloud-dotnet/tree/issue1036/issue1036

![image](https://cloud.githubusercontent.com/assets/80104/25915119/b43a5802-35c0-11e7-81c8-fa35843a9b41.png)

``` csharp
namespace issue1036
{
    class Program
    {
        const string LogId = "issue1036";

        static void Main(string[] args)
        {
            var appender = new Google.Cloud.Logging.Log4Net.GoogleStackdriverAppender();
            appender.ProjectId = "tachyusweb";
            appender.LogId = LogId;
            appender.Layout = new log4net.Layout.PatternLayout { ConversionPattern = "%message" };
            appender.ActivateOptions();

            var hierarchy = (log4net.Repository.Hierarchy.Hierarchy)log4net.LogManager.GetRepository();
            hierarchy.Root.AddAppender(appender);
            hierarchy.Root.Level = log4net.Core.Level.Debug;
            hierarchy.Configured = true;

            var log = log4net.LogManager.GetLogger(LogId);
            log.Info("a");
            log.Info("b");
            log.Info("c");
            appender.AddCustomLabel(new Google.Cloud.Logging.Log4Net.GoogleStackdriverAppender.Label { Key = "batch", Value = "1" });
            log.Info("d");
            log.Info("e");
            log.Info("f");
            appender.AddCustomLabel(new Google.Cloud.Logging.Log4Net.GoogleStackdriverAppender.Label { Key = "batch", Value = "2" });
            log.Info("g");
            log.Info("h");
            log.Info("i");

            appender.Dispose();
        }
    }
}
```

**Comment by jskeet** [comment]
Apologies for not replying before - looking now. (My guess is that it's a dictionary Add fail somewhere.)

**Comment by jskeet** [comment]
Okay, as I thought - this is the culprit, in `GoogleStackDriverAppender`:

```csharp
var labels = new Dictionary<string, string>();
foreach (var customLabel in _customLabels)
{
    labels.Add(customLabel.Key, customLabel.Value);
}
```

Note that if you'd added them both before calling `ActivateOptions`, that would have thrown.

@chrisdunelm: Is it intended that users can change the appender configuration after activating it? I can see how it would be useful to have replaceable labels like this...

(As an aside, we also have a pointless loop like the above in `ActivateOptions`, where the labels aren't actually used afterwards...)

**Comment by chrisdunelm** [comment]
It was never intended that labels should be changed after activation.
Unless there is a compelling case to allow configuration to be changed after activation, I'd prefer that the code is changed to throw on any post-activation configuration modification attempt.

**Comment by jskeet** [comment]
@chrisdunelm: So is the plan to address this issue by changing it so it *will* throw?

**Comment by chrisdunelm** [comment]
@jskeet Yes

**Comment by chrisdunelm** [comment]
Fixed by #1093
